segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/hdfs/HDFSStorage.java

```diff
@@ -196,86 +328,302 @@ public boolean supportsTruncation() {
+            try {
+                FileStatus existingFiles = findStatusForSegment(streamSegmentName, true);
+
+                if (!isSealed(existingFiles.getPath())) {
+                    if (getEpocFromPath(existingFiles.getPath()) > this.epoch) {
+                        throw new StorageNotPrimaryException(streamSegmentName);
+                    }
+
+                    Path targetPath = getFilePath(streamSegmentName, this.epoch);
+                    if (!targetPath.equals(existingFiles.getPath())) {
+                        try {
+                            this.fileSystem.rename(existingFiles.getPath(), targetPath);
+                        } catch (FileNotFoundException e) {
+                            //This happens when more than one host is trying to fence and only one of the host goes through
+                            // retry the rename so that host with the highest epoch gets access.
+                            log.warn("Race in fencing. More than two hosts trying to own the segment. Retrying");
+                            fencedCount++;
+                            continue;
+                        }
+                    }
+                }
+                //Ensure that file exists
+                findStatusForSegment(streamSegmentName, true);
+                return HDFSSegmentHandle.write(streamSegmentName);
+            } catch (IOException e) {
+                throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+            }
+        } while (fencedCount < MAX_ATTEMPT_COUNT);
+        LoggerHelpers.traceLeave(log, "openWrite", traceId, epoch);
+        return null;
     }
 
     @Override
-    @SneakyThrows(StreamSegmentException.class)
-    public boolean exists(String streamSegmentName) {
-        return call(new ExistsOperation(streamSegmentName, this.context));
-    }
+    public SegmentProperties create(String streamSegmentName) throws StreamSegmentException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+        // Create the segment using our own epoch.
+        FileStatus[] existingFiles = null;
+        try {
+            existingFiles = findAllRaw(streamSegmentName);
+        } catch (IOException e) {
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, e);
+        }
+        if (existingFiles != null && existingFiles.length > 0) {
+            // Segment already exists; don't bother with anything else.
+            throw HDFSExceptionHelpers.throwException(streamSegmentName, HDFSExceptionHelpers.segmentExistsException(streamSegmentName));
+        }
 
+        // Create the first file in the segment.
+        Path fullPath = getFilePath(streamSegmentName, this.epoch);
+        try {
+            // Create the file, and then immediately close the returned OutputStream, so that HDFS may properly create the file.
+            this.fileSystem.create(fullPath, READWRITE_PERMISSION, false, 0, this.config.getReplication(),
```

Comment:
How do you handle races here in the case when two different hosts with different epochs both made it to this point and both create files with different epochs?